src/Cardano/Wallet/CoinSelection/LargestFirst.hs

```diff
@@ -0,0 +1,112 @@
+{-# LANGUAGE RankNTypes #-}
+
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- This module contains the implementation of largestFirst
+-- input selection algorithm
+
+
+module Cardano.Wallet.CoinSelection.LargestFirst (
+    largestFirst
+  ) where
+
+import Prelude
+
+import Cardano.Wallet.CoinSelection
+    ( CoinSelection (..), CoinSelectionError (..), CoinSelectionOptions (..) )
+import Cardano.Wallet.Primitive.Types
+    ( Coin (..), TxIn, TxOut (..), UTxO (..), balance )
+import Control.Monad
+    ( foldM, when )
+import Control.Monad.IO.Class
+    ( MonadIO )
+import Control.Monad.Trans.Except
+    ( ExceptT (..), throwE )
+import Data.List.NonEmpty
+    ( NonEmpty (..) )
+import Data.Ord
+    ( comparing )
+
+import qualified Data.List as L
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map.Strict as Map
+
+
+-- | Largest-first input selection policy
+largestFirst
+    :: forall m. MonadIO m
+    => CoinSelectionOptions
+    -> UTxO
+    -> NonEmpty TxOut
+    -> ExceptT CoinSelectionError m CoinSelection
+largestFirst opt utxo txOutputs = do
+    let txOutputsSorted = NE.toList $ NE.sortBy (comparing coin) txOutputs
+    let n = fromIntegral $ maximumNumberOfInputs opt
+    let nLargest = take n . L.sortBy (flip $ comparing (coin . snd)) . Map.toList . getUTxO
+    let moneyRequested = sum $ (getCoin . coin) <$> txOutputsSorted
+    let utxoBalance = fromIntegral $ balance utxo
+    let numberOfUtxoEntries = fromIntegral $ L.length $ (Map.toList . getUTxO) utxo
+    let numberOfTransactionOutputs = fromIntegral $ NE.length txOutputs
+
+    when (numberOfUtxoEntries < numberOfTransactionOutputs)
+        $ throwE $ UtxoNotEnoughFragmented numberOfUtxoEntries numberOfTransactionOutputs
+
+    when (utxoBalance < moneyRequested)
+        $ throwE $ UtxoExhausted utxoBalance moneyRequested
```

Comment:
Also, I think that, those checks where done in the `Nothing` branch previously, such that they will only occur _after_ the coin selection has been tried once. I think the rational was that successful calls to this function are more likely than error calls and thus, rather than doing those checks on every call, we simply do it in case of error and save some CPU cycles along the way.

Reply:
swapped the errors, checked it in unit tests, and changed the name of the error